// @opencensus/exporter-stackdriver/build/src/types.js
"use strict";
/**
 * Copyright 2018, OpenCensus Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var MetricKind;
(function (MetricKind) {
    MetricKind["UNSPECIFIED"] = "METRIC_KIND_UNSPECIFIED";
    MetricKind["GAUGE"] = "GAUGE";
    MetricKind["DELTA"] = "DELTA";
    MetricKind["CUMULATIVE"] = "CUMULATIVE";
})(MetricKind = exports.MetricKind || (exports.MetricKind = {}));
var ValueType;
(function (ValueType) {
    ValueType[ValueType["VALUE_TYPE_UNSPECIFIED"] = 0] = "VALUE_TYPE_UNSPECIFIED";
    ValueType[ValueType["BOOL"] = 1] = "BOOL";
    ValueType[ValueType["INT64"] = 2] = "INT64";
    ValueType[ValueType["DOUBLE"] = 3] = "DOUBLE";
    ValueType[ValueType["STRING"] = 4] = "STRING";
    ValueType[ValueType["DISTRIBUTION"] = 5] = "DISTRIBUTION";
    ValueType[ValueType["MONEY"] = 6] = "MONEY";
})(ValueType = exports.ValueType || (exports.ValueType = {}));
